src/theory/fp/fp_converter.cpp

```diff
@@ -12,25 +12,1029 @@
 }
 
+
+FpConverter::FpConverter(context::UserContext *user)
+    : f(user), r(user), b(user), u(user), s(user),
+  d_additionalAssertions(user) {}
+
+  Node FpConverter::ufToNode (const fpt &format, const uf &u) const {
+    NodeManager *nm = NodeManager::currentNM();
+
+    FloatingPointSize fps(format.getTypeNode().getConst<FloatingPointSize>());
+
+    // This is not entirely obvious but it builds a float from components
+    // Particularly, if the components can be constant folded, it should
+    // build a Node containing a constant FloatingPoint number
+
+    ubv packed(symfpu::pack<traits>(format, u));
+    Node value = nm->mkNode(nm->mkConst(FloatingPointToFPIEEEBitVector(fps)),
+			    packed.getNode());
+    return value;
+  }
+
+  Node FpConverter::rmToNode (const rm &r) const {
+    NodeManager *nm = NodeManager::currentNM();
+
+    Node transVar = r.getNode();
+
+    Node RNE = traits::RNE().getNode();
+    Node RNA = traits::RNA().getNode();
+    Node RTP = traits::RTP().getNode();
+    Node RTN = traits::RTN().getNode();
+    Node RTZ = traits::RTZ().getNode();
+
+    Node value =
+      nm->mkNode(kind::ITE,
+		 nm->mkNode(kind::EQUAL, transVar, RNE),
+		 nm->mkConst(roundNearestTiesToEven),
+		 nm->mkNode(kind::ITE,
+			    nm->mkNode(kind::EQUAL, transVar, RNA),
+			    nm->mkConst(roundNearestTiesToAway),
+			    nm->mkNode(kind::ITE,
+				       nm->mkNode(kind::EQUAL, transVar, RTP),
+				       nm->mkConst(roundTowardPositive),
+				       nm->mkNode(kind::ITE,
+						  nm->mkNode(kind::EQUAL, transVar, RTN),
+						  nm->mkConst(roundTowardNegative),
+						  nm->mkConst(roundTowardZero)))));
+    return value;
+  }
+
+  Node FpConverter::propToNode (const prop &p) const {
+    NodeManager *nm = NodeManager::currentNM();
+    Node value = nm->mkNode(kind::EQUAL,
+			    p.getNode(),
+			    nm->mkConst(::CVC4::BitVector(1U, 1U)));
+    return value;
+  }
+  Node FpConverter::ubvToNode (const ubv &u) const {
+    return u.getNode();
+  }
+  Node FpConverter::sbvToNode (const sbv &s) const {
+    return s.getNode();
+  }
+
+
+  // Creates the components constraint
+  FpConverter::uf FpConverter::buildComponents(TNode current) {
+    Assert(Theory::isLeafOf(current, THEORY_FP) ||
+	   current.getKind() == kind::FLOATINGPOINT_TO_FP_REAL);
+
+
+    NodeManager *nm = NodeManager::currentNM();
+    uf tmp(nm->mkNode(kind::FLOATINGPOINT_COMPONENT_NAN, current),
+	   nm->mkNode(kind::FLOATINGPOINT_COMPONENT_INF, current),
+	   nm->mkNode(kind::FLOATINGPOINT_COMPONENT_ZERO, current),
+	   nm->mkNode(kind::FLOATINGPOINT_COMPONENT_SIGN, current),
+	   nm->mkNode(kind::FLOATINGPOINT_COMPONENT_EXPONENT, current),
+	   nm->mkNode(kind::FLOATINGPOINT_COMPONENT_SIGNIFICAND, current));
+
+    d_additionalAssertions.push_back(tmp.valid(fpt(current.getType())).getNode());
+
+    return tmp;
+  }
+
+
+
+  // Non-convertible things should only be added to the stack at the very start, thus...
+  #define PASSTHROUGH   Assert(workStack.empty())
```

Comment:
Macros should start with `CVC4_`, see https://github.com/CVC4/CVC4/wiki/Code-Style-Guidelines#naming-macros.

Reply:
Done.